### PR TITLE
Add a container terminal to the pod details page

### PR DESCRIPTION
**Pull request by stefwalter**
Add a container terminal to the pod details page
    
This uses the container terminal widget:
    
https://github.com/kubernetes-ui/container-terminal/

### CODE REVIEW

**Comment by stefwalter** [comment]
This commit depends on #4986 and has been cherry-picked into this branch.

**Comment by mfojtik** [comment]
@stefwalter screenshots! :-)

**Comment by stefwalter** [comment]
@mfojtik I'll do a little video, just looking for some good pod definitions.

**Comment by stefwalter** [comment]
I'm having problems with the minification here, will work on this a bit more before its ready.

**Comment by stefwalter** [comment]
I think #4873 broke WebSocket access to containers (at least for me).

**Comment by stefwalter** [comment]
@jwforres This is ready for review. I won't rebase until the issue with #4873 is resolved.

**Comment by stefwalter** [comment]
[Fixed behavior](http://www.youtube.com/watch?v=SMxVQBD3Kho)]

**Comment by liggitt** [comment]
I think you just need to reconcile roles and role bindings to allow the new auth chain checks from the kubelet

**Comment by jwforres** [comment]
did an initial review of the asset changes, would like @liggit to look at the change to auth.js

rebase should fix the diff for all that annotation stuff

**Comment by jwforres** [comment]
@stefwalter you will also want to wait for this to merge before rebasing / rebuilding bindata: https://github.com/openshift/origin/pull/5023   we are having merge conflicts on minified templates and this should alleviate a lot of those.

File: assets/app/scripts/services/auth.js [hunk 1]
```diff
@@ -276,4 +276,18 @@ angular.module('openshiftConsole')
       }
     }
   };
-}]);
+}])
+
+/* A WebSocket factory for kubernetesContainerTerminal */
+.factory("AuthContainerWebSocket", function(API_CFG, AuthService) {
```
Comment (liggitt) [inline assets/app/scripts/services/auth.js]: why not use $ws?

Comment (stefwalter) [inline assets/app/scripts/services/auth.js]: Because the model does not fit with what container-terminal needs. It needs a factory of WebSocket() like objects. $ws returns promises. I'd rather not export that complexity to other projects using the container-terminal widget.

File: assets/app/views/browse/pod.html [hunk 2]
```diff
@@ -86,6 +86,22 @@
               </tab>
               -->
 
+              <tab heading="Terminal" active="selectedTab.terminal"
+                select="renderOptions.preventContainerSockets = false"
```
Comment (liggitt) [inline assets/app/views/browse/pod.html]: renderOptions seems like an odd place to put this... what is this doing?

Comment (jwforres) [inline assets/app/views/browse/pod.html]: Yeah renderOptions is usually reserved for things that effect what renders
in the chrome around the main content area, like the label filter widget.
You should just set this on something like
$scope.terminal.preventContainerSockets

@liggit this is preventing the websocket from connecting until the first
time you open the tab, but then allows it to remain open as long as you
just switch between the tabs on the pod so commands can keep running. It
only kills the connection if you leave the pod page.

File: assets/app/scripts/services/auth.js (hunk 1 above)
Comment (liggitt) [inline assets/app/scripts/services/auth.js]: Allow the web socket factory to return a web socket or the promise of a web socket, and no complexity is added to other projects. We should allow for the case where work is needed to go get tokens, etc, to establish the web socket call.

Comment (liggitt) [inline assets/app/scripts/services/auth.js]: Also, `$ws` should be updated to plumb the optional second `protocols` param to `new WebSocket`

Comment (liggitt) [inline assets/app/scripts/services/auth.js]: I want to consistently use $ws and $http to make sure registered interceptors get called in all cases

File: assets/app/views/browse/pod.html (hunk 2 above)
Comment (liggitt) [inline assets/app/views/browse/pod.html]: oh... that's a little hard to follow. could we do this instead:
`select="terminalTabSelected = true"`
...
`prevent="!terminalTabSelected && !selectedTab.terminal"`

File: assets/app/scripts/services/auth.js (hunk 1 above)
Comment (stefwalter) [inline assets/app/scripts/services/auth.js]: Done.

File: assets/app/views/browse/pod.html (hunk 2 above)
Comment (stefwalter) [inline assets/app/views/browse/pod.html]: Changed

**Comment by stefwalter** [comment]
> you will also want to wait for this to merge before rebasing / rebuilding bindata: #5023 we are having merge conflicts on minified templates and this should alleviate a lot of those.

There have always been lots of merge conflicts. Happy to hear it's being resolved ... however I don't think that's a blocker to getting this in.

File: assets/app/scripts/services/ws.js [hunk 3]
```diff
@@ -30,13 +30,13 @@ angular.module('openshiftConsole')
...
     // Implement $ws()
-    var $ws = function(config) {
+    var $ws = function(config, protocols) {
       config.method = angular.uppercase(config.method || "WATCH");
 
       authLogger.log("$ws (pre-intercept)", config.url.toString());
       var serverRequest = function(config) {
         authLogger.log("$ws (post-intercept)", config.url.toString());
-        var ws = new WebSocket(config.url);
+        var ws = new WebSocket(config.url, protocols);
```
Comment (liggitt) [inline assets/app/scripts/services/ws.js]: can you put protocols on the config object, so `config.url, config.protocols`?

**Comment by jwforres** [comment]
If any web console PRs merge then we will have to rebuild #5023 so we
aren't merging any until it goes in.  It won't block it for long, we are
merging it today.

File: assets/app/scripts/services/auth.js [hunk 4]
```diff
@@ -276,4 +276,16 @@ angular.module('openshiftConsole')
       }
     }
   };
-}]);
+}])
+
+/* A WebSocket factory for kubernetesContainerTerminal */
+.factory("AuthContainerWebSocket", function(API_CFG, $ws) {
```
Comment (liggitt) [inline assets/app/scripts/services/auth.js]: because we're getting auth by virtue of the registered interceptors now, this is orthogonal to auth... move to ws.js or to its own file?

Comment (stefwalter) [inline assets/app/scripts/services/auth.js]: Done.

File: assets/app/scripts/services/ws.js (hunk 3 above)
Comment (stefwalter) [inline assets/app/scripts/services/ws.js]: Done.

**Comment by stefwalter** [comment]
Rebased and fixed remaining issues.

File: assets/app/scripts/services/ws.js [hunk 5]
```diff
@@ -75,4 +75,16 @@ angular.module('openshiftConsole')
 
     return $ws;
   };
+})
+
+/* A WebSocket factory for kubernetesContainerTerminal */
+.factory("ContainerWebSocket", function(API_CFG, $ws) {
```
Comment (liggitt) [inline assets/app/scripts/services/ws.js]: @spadgett fyi with your auth changes... we might want this in data.js eventually

**Comment by liggitt** [comment]
auth/ws changes LGTM

**Comment by jwforres** [comment]
@stefwalter is this ready for QE to test, if so we should create a fork_ami

**Comment by stefwalter** [comment]
Looks like it needs another rebase ... should be ready in half an hour (after install-assets.sh completes again).

**Comment by jwforres** [comment]
@stefwalter yeah i typically skip install-assets when i can, you should only need that if a dependency update has occurred, you can check bower.json/package.json to see if there have been any updates

**Comment by stefwalter** [comment]
> yeah i typically skip install-assets when i can, you should only need that if a dependency update has occurred, you can check bower.json/package.json to see if there have been any updates

This pull requests adds a dependency, so I find i have to run install-assets.sh or bindata.go becomes very messed up.

**Comment by stefwalter** [comment]
Rebased, and ready for testing.

**Comment by jwforres** [comment]
@stefwalter ok I kicked off a fork_ami for it https://ci.openshift.redhat.com/jenkins/job/fork_ami/99/

**Comment by jwforres** [comment]
@stefwalter its failing on the ui e2e tests https://ci.openshift.redhat.com/jenkins/job/fork_ami/100/console

**Comment by jwforres** [comment]
@stefwalter the fork ami for QE to test can't complete unless the e2e tests are passing

**Comment by spadgett** [comment]
@stefwalter update karma.conf.js for the new dependency?

**Comment by stefwalter** [comment]
Added to karma.conf.js and Rebased.

**Comment by liggitt** [comment]
three quick things that occurred to me, just wanted to make sure we're handling:
* tearing down the web socket on scope destroy
* user is not permitted to exec and gets a 403 (project viewer, etc)
* web socket timeout/reestablishment (either automatic or a "reconnect"-type UI)

If there are multiple containers, are we automatically establishing terminals to all of them? I'm a little concerned about that in cases of large numbers of containers

**Comment by stefwalter** [comment]
> tearing down the web socket on scope destroy

Yup: https://github.com/kubernetes-ui/container-terminal/blob/master/container-terminal.js#L218

> user is not permitted to exec and gets a 403 (project viewer, etc)

Yup. We show a generic message. The WebSocket API sucks and doesn't give further info. This was discussed during design review.

https://github.com/kubernetes-ui/container-terminal/blob/master/container-terminal.js#L144

> web socket timeout/reestablishment (either automatic or a "reconnect"-type UI)

First of all there's a ping that keeps it alive. Secondly, upon disconnect a button appears. See the video.

> If there are multiple containers, are we automatically establishing terminals to all of them? I'm a little concerned about that in cases of large numbers of containers

There's no such check. I think one could be added later if it proves to be a problem.

**Comment by liggitt** [comment]
Oops, missed the video...

**Comment by stefwalter** [comment]
Rebased ... and also:
 * Add 'fa' class to fix terminal issues
 * Don't show 'Terminal' tab if there's no running containers.

**Comment by jwforres** [comment]
[merge]

**Comment by jwforres** [comment]
@stefwalter lot of stuff merging today, something caused a merge conflict

File: assets/app/scripts/services/ws.js [hunk 6]
```diff
@@ -75,4 +75,16 @@ angular.module('openshiftConsole')
...
+  return function AuthWebSocket(url, protocols) {
+    var scheme;
+    if (url.indexOf("/") === 0) {
+      scheme = window.location.protocol === "http:" ? "ws://" : "wss://";
+      url = scheme + API_CFG.openshift.hostPort + url;
+    }
+    return $ws({ url: url, method: "WATCH", protocols: protocols });
```
Comment (liggitt) [inline assets/app/scripts/services/ws.js]: to automatically pick up auth token, you need to set an `auth` stanza in the options, like `auth:{}` (data.js adds this for any `$http`/`$ws` requests it generates

Comment (stefwalter) [inline assets/app/scripts/services/ws.js]: Thanks. Done.

**Comment by stefwalter** [comment]
Rebased.

**Comment by jwforres** [comment]
[merge]

**Comment by jwforres** [comment]
geez another merge conflict :-(  @stefwalter

**Comment by stefwalter** [comment]
Rebased. The first commit got unceremoniously squashed into 1a48541fb3b8a3253d1daf1b23d2e4e7918e54d6 ... so just one left.

**Comment by jwforres** [comment]
[merge]
